### DISCUSSION TITLE
docs: How to use the seeder, generator, etc. with ts-node in a project using node's native ECMAScript module loader

### DISCUSSION THREAD

**Discussion by camflint** [mikro-orm/mikro-orm#4072] [Ideas]
It took me quite a while to get this particular setup working, so I wanted to share what worked for me in hopes that it might be useful, and maybe even make its way into the docs.

First, the relevant bits of configuration (some obviously omitted for brevity):

package.json
```json
  "mikro-orm": {
    "useTsNode": true,
    "configPaths": [
      "./src/mikro-orm.config.ts"
    ]
  },
  "dependencies": {
    "ts-node": "^10.9.1",
    "ts-paths-esm-loader": "^1.4.1"
  },
  "scripts": {
    "db:seeder:run": "NODE_OPTIONS='--loader ts-paths-esm-loader/transpile-only --no-warnings' TS_NODE_PREFER_TS_EXTS=true npx mikro-orm seeder:run"
  }
```

src/mikro-orm.config.ts
```ts
import { defineConfig } from '@mikro-orm/core';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';
import { SqliteDriver } from '@mikro-orm/sqlite';

function dirname() {
  return path.dirname(url.fileURLToPath(import.meta.url));
}

export default defineConfig<SqliteDriver>({
  metadataProvider: TsMorphMetadataProvider,
  schemaGenerator: {
    disableForeignKeys: true,
    createForeignKeyConstraints: true,
    ignoreSchema: [],
  },
  migrations: {
    path: dirname() + '/migrations',
    pathTs: dirname() + '/migrations',
    tableName: 'migrations',
  },
  seeder: {
    path: dirname() + '/seeders',
    pathTs: dirname() + '/seeders',
  },
  entities: [dirname() + '/entities/**/*.js'],
  entitiesTs: [dirname() + '/entities/**/*.ts'],
  type: 'sqlite',
  forceUtcTimezone: false,
  dbName: 'my-db-name',
  debug: true, // false for production
});
```

tsconfig.json
```json
{
  "include": ["src/**/*"],
  "compilerOptions": {
    "target": "ES2022",
    "lib": ["ES2022"],

    // Modules.
    "module": "NodeNext",
    "moduleResolution": "nodenext",

    // Path mapping.
    "baseUrl": ".",

    // Emit.
    "declaration": true,
    "sourceMap": true,
    "emitDecoratorMetadata": true,

    // Interop constraints.
    "forceConsistentCasingInFileNames": true,

    // Decorators.
    "experimentalDecorators": true,

    // Type checking.
    "strict": true,
  },
  "ts-node": {
    // (don't need either of these when using ts-paths-esm-loader)
    //"esm": true,
    //"experimentalResolver": true
  }
}
```

The key parts are:
1. Using `ts-paths-esm-loader/transpile-only` as the custom loader. `ts-node/esm` did NOT work because it couldn't handle `import ... 'foo.js'` (import statements with generated-file extensions).
2. Use of `TS_NODE_PREFER_TS_EXTS=true` flag. Without this it seems like Mikro would load inconsistent entity hashes (see error trace below).

This is the kind of error I was running into before I landing on the config above:

```
ValidationError: Trying to persist not discovered entity of type Category. Entity with this name was discovered, but not the prototype you are passing to the ORM. If using EntitySchema, be sure to point to the implementation via `class`.
    at Function.notDiscoveredEntity (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/core/errors.js:50:16)
    at SqlEntityManager.persist (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/core/EntityManager.js:888:48)
    at CategoriesSeeder.run (file:///Users/cameron/Projects/personalorganizer-js/src/seeders/CategoriesSeeder.ts:266:10)
    at DatabaseSeeder.call (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/seeder/Seeder.js:9:28)
    at DatabaseSeeder.run (file:///Users/cameron/Projects/personalorganizer-js/src/seeders/DatabaseSeeder.ts:8:17)
    at SeedManager.seed (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/seeder/SeedManager.js:27:26)
    at SeedManager.seedString (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/seeder/SeedManager.js:50:24)
    at async Object.handler (/Users/cameron/Projects/personalorganizer-js/node_modules/@mikro-orm/cli/commands/DatabaseSeedCommand.js:25:9) {
  entity: undefined
}
```

Now I can just run this and it works as expected:

```
npm run db:seeder:run
```

**Comment by B4nan** [comment]
Thanks for the write-up, but I already have basic ESM setup covered in the new getting started guide I am working on (should be out together with v6), so closing. For me it was working just fine with much simpler setup, e.g. no need to override any of the ORM settings, no need for absolute paths at all, and I can also say I never had to use `TS_NODE_PREFER_TS_EXTS`.

> This is the kind of error I was running into before I landing on the config above:

This means you have multiple classes in the context, with the same name. Usually points to using both JS and TS imports.

Looks like the main thing you missed is there is `mikro-orm-esm` binary you need to use in ESM projects. So instead of `NODE_OPTIONS='--loader ts-paths-esm-loader/transpile-only --no-warnings' TS_NODE_PREFER_TS_EXTS=true npx mikro-orm seeder:run` you should be doing just `npx mikro-orm-esm seeder:run`, will add a note to the docs about that, as it is indeed not there, I was hoping to have the new getting started guide ready much sooner...

**Comment by ddadaal** [comment]
I encoutered this when migrating my project to esm, and the original post helps me a lot. 

The `ts-paths-esm-loader` is needed to support path mapping (`import {} from "src/config.js";` for example), because `tsconfig-paths` doesn't work with custom ES module loader. `mikro-orm-esm` doesn't support path mapping or work with the `ts-paths-esm-loader` loader either.

**Reply by B4nan** [reply]
I see, so the issue is with paths mapping. I will experiment with this a bit to see if we can support it with the `mikro-orm-esm` command out of box, if you know how, PR welcome!

edit: but if it requires to modify the binary, we would probably need yet another binary, as the `ts-paths-esm-loader` dependency needs to stay optional
